python/paddle/fluid/tests/unittests/test_batch_norm_mkldnn_op.py

```diff
@@ -0,0 +1,163 @@
+                        "Mean": block.var('mean'),
+                        "Variance": block.var('variance')
+                    },
+                    outputs={
+                        "Y": block.var('y'),
+                        "MeanOut": block.var('mean'),  # share the same memory
+                        "VarianceOut":
+                        block.var('variance'),  # share the same memory
+                        "SavedMean": block.var('saved_mean'),
+                        "SavedVariance": block.var('saved_variance')
+                    },
+                    attrs={
+                        "momentum": momentum,
+                        "epsilon": epsilon,
+                        "is_test": False,
+                        "data_layout": data_layout,
+                        "use_mkldnn": self.use_mkldnn
+                    })
+                block.create_var(name='y@GRAD', dtype='float32', shape=y.shape)
+
+                # generate backward op_desc
+                grad_op_desc_list, op_grad_to_var = core.get_grad_op_desc(
+                    bn_op.desc, set(), [])
+                grad_op_desc = grad_op_desc_list[0]
+                new_op_desc = block.desc.append_op()
+                new_op_desc.copy_from(grad_op_desc)
+                for var_name in grad_op_desc.output_arg_names():
+                    block.desc.var(var_name.encode("ascii"))
+                grad_op_desc.infer_var_type(block.desc)
+                grad_op_desc.infer_shape(block.desc)
+                for arg in grad_op_desc.output_arg_names():
+                    grad_var = block.desc.find_var(arg.encode("ascii"))
+                    grad_var.set_dtype(core.VarDesc.VarType.FP32)
+
+                exe = fluid.Executor(place)
+                out = exe.run(
+                    program,
+                    feed={
+                        name: var_dict[name]
+                        for name in
+                        ['x', 'scale', 'bias', 'mean', 'variance', 'y@GRAD']
+                    },
+                    fetch_list=[
+                        'y', 'mean', 'variance', 'saved_mean', 'saved_variance',
+                        'x@GRAD', 'scale@GRAD', 'bias@GRAD'
+                    ])
+
+            self.__assert_close(y, out[0], "y")
+            self.__assert_close(mean_out, out[1], "mean")
+            self.__assert_close(variance_out, out[2], "variance", 1e-3)
+            self.__assert_close(saved_mean, out[3], "saved_mean")
+            self.__assert_close(saved_variance, out[4], "saved_variance", 1e-3)
+            self.__assert_close(x_grad, out[5], "x_grad")
+            self.__assert_close(scale_grad, out[6], "scale_grad")
+            self.__assert_close(bias_grad, out[7], "bias_grad")
+
+            print "op test forward backward passed: ", str(place), data_layout
+
+        place = core.CPUPlace()
+        data_format = "NCHW"
+        test_with_place(place, data_format, [2, 3, 4, 5])
```

Comment:
@tpatejko I see the only difference between `test_with_place` of `test_batch_norm_mkldnn_op.py` and `test_batch_norm_op.py` is line 146-148:
```
+        place = core.CPUPlace()
+        data_format = "NCHW"
+        test_with_place(place, data_format, [2, 3, 4, 5])
```
and
https://github.com/PaddlePaddle/Paddle/blob/ad91bfe5d4531c1a95dc968bed25ad99a0cf779f/python/paddle/fluid/tests/unittests/test_batch_norm_op.py#L392-L398

Thus, how about move `test_with_place` out of `test_forward_backward` in `test_batch_norm_op.py` like:
```
def test_with_place(self, place, data_layout, shape):
   ....
def test_forward_backward(self):
        places = [core.CPUPlace()]
        if core.is_compiled_with_cuda() and core.op_support_gpu("batch_norm"):
            places.append(core.CUDAPlace(0))

        for place in places:
            for data_format in ["NCHW", "NHWC"]:
                self.test_with_place(place, data_format, [2, 3, 4, 5])
``` 
Then, you can only rewrite `test_forward_backward` in `test_batch_norm_mkldnn_op.py`, likes:
```
class TestMKLDNNBatchNormOpTraining(TestBatchNormOpTraining):
  def init_kernel_type(self):
       self.use_mkldnn = True
  def test_forward_backward(self):
       place = core.CPUPlace()
       data_format = "NCHW"
       self. test_with_place(place, data_format, [2, 3, 4, 5])
```

Reply:
@luotao1 One more thing: you are right about moving `test_with_place` out of `test_backward_forward` and adding data format as a parameter. I did it, but the test cases for MKLDNN were failing because of inverted batch/saved variance.

Reply:
@tpatejko Thanks for your explanation. I see it. But how about add a func `_reference_training_and_grad` in  `TestBatchNormOpTraining` likes:
```
def _reference_training_and_grad(x, scale, bias, epsilon, data_layout):
           y, saved_mean, saved_variance = _reference_training(
                x, scale, bias, epsilon, data_layout)
            mean_out = saved_mean * (1. - momentum) + momentum * mean
            variance_out = saved_variance * (1. - momentum
                                             ) + momentum * variance
            # run backward
            y_grad = np.random.random_sample(shape).astype(np.float32)
            x_grad, scale_grad, bias_grad = _reference_grad(
                x, y_grad, scale, saved_mean, saved_variance, epsilon,
                data_layout)
            return x_grad, scale_grad, bias_grad
```
Then, you can only rewrite `_reference_training_and_grad` in `test_batch_norm_mkldnn_op.py`.
The reason is:
- There are a lot of similar codes in `test_batch_norm_mkldnn_op.py` and `test_batch_norm_op.py`.
- It's hard to find out the different MKLDNN implementation of batch normalization which does not invert batch variance (saved variance).

Reply:
@luotao1 I think doing reference training in a separate function is a good idea. I will try to implement it.